[Chrysline Lim] iP

ChryslineLim opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use 

All you need to do is, 

1. download it from [here](https://github.com/nus-cs2103-AY2223S1/ip/pull/432).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) 

---

If you are a Java programmer, you can use it to practise Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ChryslineLim committed:
Level-1: Greet, Echo, Exit

ChryslineLim committed:
Level-2: Add, List

ChryslineLim committed:
Level-3: Mark as Done

ChryslineLim committed:
Level-4: ToDo, Event, Deadline

ChryslineLim committed:
A-TextUiTesting: Automated Text UI Testing

ChryslineLim committed:
Level-5: Handle Errors

ChryslineLim committed:
Level-6: Delete

ChryslineLim committed:
Add Level-7 : Save

ChryslineLim committed:
Add Level-8: Dates and Times

ChryslineLim committed:
Merge branch 'branch-Level-7'

ChryslineLim committed:
Resolve merge conflict

ChryslineLim committed:
Convert deadline date saved in data file into LocalDate object

ChryslineLim committed:
Add A-MoreOOP: Use More OOP

ChryslineLim committed:
Fix todo description cutting off bug

ChryslineLim committed:
Add A-Packages: Organize into Packages

ChryslineLim committed:
Merge branch 'add-gradle-support' of https://github.com/nus-cs2103-AY2223S1/ip

ChryslineLim committed:
Add A-JUnit: Add JUnit Tests

ChryslineLim committed:
Tweak A-CodingStandard

ChryslineLim committed:
Add Level-9: Find

ChryslineLim committed:
Add A-JavaDoc: JavaDoc

ChryslineLim committed:
Merge branch 'branch-Level-9'

ChryslineLim committed:
Merge branch 'branch-A-CodingStandard'

ChryslineLim committed:
Merge branch 'branch-A-JavaDoc'

ChryslineLim committed:
change boolean name

ChryslineLim committed:
remove unnecessary import

shenyih0ng commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,61 @@
+import duke.task.Task;
+import duke.task.ToDoTask;
+
+import java.time.LocalDate;
```
Standard `java.` imports should be placed above `duke.` imports according to `checkstyle` rules

shenyih0ng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,56 @@
+            return new ListCommand();
+        }
+        
+        if (!command.equals("mark") && !command.equals("unmark") && !command.equals("todo") && !command.equals("deadline") && !command.equals("event") && !command.equals("delete") && !command.equals("find")) {
+            throw new DukeException("OOPS!!! I'm sorry, but I don't know what that means :-(");
+        }
```
Instead of doing a check for unknown commands here, maybe it will be clearer if you shift this integrate this check into the switch statement below (L33-L55).

For example:

```java
switch (command) {
.
.
.
default:
    throw new DukeException("...");
}
```

This will remove the need to write long condition checks as the number of command type increase.

ChryslineLim replied:
I placed it on top in order to be before the "OOPS!!! The description cannot be empty." error, otherwise it would print that for any unknown commands of only one word :( I also couldn't shift the empty error to the end as I needed to ensure the commands have a description before the switch cases :(

shenyih0ng commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,56 @@
+        String suffix = spacedArr[1];
+
+        switch (command) {
+        case "mark": {
+            int pos = Integer.parseInt(suffix) - 1;
+            return new MarkCommand(true, pos);
+        }
+        case "unmark": {
+            int pos = Integer.parseInt(suffix) - 1;
+            return new MarkCommand(false, pos);
+        }
+        case "todo":
+            return new AddCommand("todo", suffix);
+        case "deadline":
+            return new AddCommand("deadline", suffix);
+        case "event":
+            return new AddCommand("event", suffix); 
+        case "find": 
+            return new FindCommand(suffix);
+        default: {
+            int pos = Integer.parseInt(suffix) - 1;
+            return new DeleteCommand(pos);
```
Awesome job following the correct style for switch case statements 👍

shenyih0ng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+
+
+public class Storage {
+    private final String FILE_PATH;
+    
+    public Storage(String FILE_PATH) {
```
Great job at setting `FILE_PATH` as `final`!

However, I am not quite sure if `FILE_PATH` in this case can be considered a constant since you are initializing it in the constructor. Perhaps the naming convention in the case should not be that of a constant.

shenyih0ng commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+     * @throws DukeException If data file cannot be accessed.
+     */
+    public ArrayList<Task> load() throws DukeException {
+        ArrayList<Task> data = new ArrayList<>();
```
Instead of naming it as `data`, I would think it will be semantically clearer if it is named something like `tasks`, since it is meant to represent a collection of `Task`.

shenyih0ng commented on src/test/java/duke/command/TaskListStub.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.util.ArrayList;
+
+public class TaskListStub extends TaskList {
```
Great use of a stub for `TaskList` testing 👍

shenyih0ng commented on build.gradle:
```diff
@@ -0,0 +1,41 @@
+}
+
+application {
+    mainClassName = "seedu.duke.Duke"
```
cmiiw, i believe the `mainClassName` should be `duke.Duke` instead since the your `Duke.java` is under the `duke` package

ChryslineLim replied:
Thank you for the review! I will make the necessary changes :)

shenyih0ng reviewed (commented):
Overall the coding standards have larged been adhered to! There is also clear documentation and sufficient unit tests :) 

Perhaps you can also consider adding documentation for the remaining methods/classes in the future!

ChryslineLim committed:
Add Level-10: GUI

ChryslineLim committed:
Merge branch 'branch-Level-10'

YZTangent commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,25 @@
+
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Great job naming booleans in the _isSomething_ format!

YZTangent replied:
Great job naming booleans in the isSomething format!

YZTangent commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+                    break;
+                case 'D':
+                    String[] parts = s.substring(7).split(" \\(by: ");
+                    DateTimeFormatter pattern = new DateTimeFormatterBuilder()
+                            // case-insensitive to parse JAN and FEB
+                            .parseCaseInsensitive()
+                            // add pattern
+                            .appendPattern("MMM dd yyyy")
+                            // create formatter (use English Locale to parse month names)
+                            .toFormatter(Locale.ENGLISH);
```
Great job using higher level line breaks by breaking at every operation called on DateTimeFormatterBuilder!

YZTangent commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,25 @@
+public class ParserTest {
+    
+    @Test
+    public void parse_todoBorrowBook_newAddCommand() throws DukeException {
```
Great job naming test methods using underscores and following the three part format!

ChryslineLim replied:
Thank you for the review! :)

YZTangent reviewed (commented):
The PR looks great! :) Good job following most of the coding conventions, such as casing and naming methods with verbs!

ChryslineLim committed:
Add A-Assertions

ChryslineLim committed:
Change variable names

ChryslineLim committed:
Add A-CodeQuality

Storage class methods are very long, especially for load() and writeFile().

This affects the readability and thus code quality. The implementation of writeFile()
is also unnecessarily complicated while not improving efficiency.

Let's extract methods from the load() method to break it up into smaller pieces.
Let's also change the implementation of writeFile() to be simpler.

This makes load() easier to understand by applying abstraction.

Other code refactoring was also done to improve the overall code quality.

ChryslineLim committed:
Add C-Sort

The user does not have the ability to sort their tasks by their deadlines.

This should be an important feature of a task-management app, in order to
enable users to quickly prioritise their to-do list.

Let's create a Sort command that prints to the user a list of their
tasks sorted in chronological order of deadlines. This list is a copy
of the original list, and does not modify the original list. All
tasks without deadlines will be placed below Deadline Tasks.

The original list should not be modified as the user may want it
sequenced in the order of addition of tasks. All tasks without deadlines
are de-prioritised as they are probably less urgent (and there
is no way to meaningfully sort tasks that have no deadlines).

ChryslineLim committed:
resolve merge conflicts

ChryslineLim committed:
Merge pull request #2 from ChryslineLim/branch-A-CodeQuality

A-CodeQuality

ChryslineLim committed:
resolve merge conflicts

ChryslineLim committed:
Merge pull request #3 from ChryslineLim/branch-C-Sort

Add C-Sort

ChryslineLim committed:
add back missing sort command into parser

ChryslineLim committed:
clean up imports

ChryslineLim committed:
Add some personality

ChryslineLim committed:
Improve code quality further, add more OOP, run check style

ChryslineLim committed:
fix gui bug

ChryslineLim committed:
fix file not found when launching jar bug

ChryslineLim committed:
Update README.md

ChryslineLim committed:
Remove trailing white space

ChryslineLim committed:
Merge branch 'master' of https://github.com/ChryslineLim/ip

ChryslineLim committed:
Update README.md

ChryslineLim committed:
Update README.md

ChryslineLim committed:
update docs product screenshot

ChryslineLim committed:
Merge branch 'master' of https://github.com/ChryslineLim/ip
